presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PushTopNThroughOuterJoinRuleSet.java

```diff
@@ -0,0 +1,220 @@
+
+            Optional<JoinNode> joinNodeWithTopNSource = joinNodeWithTopNSource(mappedTopN, joinNode, context);
+            return joinNodeWithTopNSource
+                    .map(newJoinNode ->
+                            Result.ofPlanNode(
+                                    parent.replaceChildren(ImmutableList.of(
+                                            projectNode.replaceChildren(ImmutableList.of(
+                                                    newJoinNode))))))
+                    .orElseGet(Result::empty);
+        }
+
+        private Optional<SymbolMapper> symbolMapper(List<Symbol> symbols, Assignments assignments)
+        {
+            SymbolMapper.Builder mapper = SymbolMapper.builder();
+            for (Symbol symbol : symbols) {
+                Expression expression = assignments.get(symbol);
+                if (!(expression instanceof SymbolReference)) {
+                    return Optional.empty();
+                }
+                mapper.put(symbol, Symbol.from(expression));
+            }
+            return Optional.of(mapper.build());
+        }
+    }
+
+    private static class PushTopNThroughOuterJoinWithoutProject
+            implements Rule<TopNNode>
+    {
+        private static final Capture<JoinNode> JOIN_CHILD = newCapture();
+
+        private static final Pattern<TopNNode> PATTERN_WITHOUT_PROJECT =
+                topN().with(source().matching(
+                        join().capturedAs(JOIN_CHILD).with(type().matching(type -> isLeftOrFullOuter(type) || isRightOrFullOuter(type)))));
+
+        @Override
+        public Pattern<TopNNode> getPattern()
+        {
+            return PATTERN_WITHOUT_PROJECT;
+        }
+
+        @Override
+        public Result apply(TopNNode parent, Captures captures, Context context)
+        {
+            JoinNode joinNode = captures.get(JOIN_CHILD);
+            TopNNode pushdownTopN = new TopNNode(context.getIdAllocator().getNextId(), joinNode, parent.getCount(), parent.getOrderingScheme(), PARTIAL);
+
+            Optional<JoinNode> joinNodeWithTopNSource = joinNodeWithTopNSource(pushdownTopN, joinNode, context);
+            return joinNodeWithTopNSource
+                    .map(newJoinNode ->
+                            Result.ofPlanNode(
+                                    parent.replaceChildren(ImmutableList.of(
+                                            newJoinNode))))
+                    .orElseGet(Result::empty);
+        }
+    }
+
+    private static Optional<JoinNode> joinNodeWithTopNSource(TopNNode topNNode, JoinNode joinNode, Context context)
+    {
+        List<Symbol> orderBySymbols = topNNode.getOrderingScheme().getOrderBy();
+        PlanNode left = joinNode.getLeft();
+        PlanNode right = joinNode.getRight();
+
+        if (isLeftOrFullOuter(joinNode.getType())
+                && ImmutableSet.copyOf(left.getOutputSymbols()).containsAll(orderBySymbols)
```

Comment:
`ImmutableSet.copyOf(left.getOutputSymbols()).containsAll(orderBySymbols)` please extract this as method  `containsAll((left.getOutputSymbols(), orderBySymbols)`. 
Or even `canSortWith(topeNNode, left)`